hashintel/hash
# Upgrade Next.js and related packages

## 🌟 What is the purpose of this PR?

The motivation for this PR is to replace webpack 4 with webpack 5 in HASH frontend (Next.js app). This unblocks @Alfred-Mountfield in https://github.com/hashintel/hash/pull/316. We can’t fix CI there because using webpack 5 in a new package causes some odd glitch between webpack 4 and webpack 5 and breaks Next.js.

This PR gives us a new version of Next.js, upgrades some Next-related dependencies and provides a slightly cleaner frontend setup.




## ⚠️ Known issues

- We now have webpack4 in blocks and webpack5 in Next.js. This would be fine if we did not `require("html-webpack-plugin")` in `next.config.js` for building block sanbox. Now that we have imports of `html-webpack-plugin` 4 and 5, `check-dependency-version-consistency` [complains](https://github.com/hashintel/hash/runs/5223019770?check_suite_focus=true).  I used `"html-webpack-plugin-v4": "npm:html-webpack-plugin@4.5.1"` in block `package.json` files, which fixes this issue without adding any special ignore rules. We will hopefully update block templates soon, so the problem will self-resolve. We need to do that anyway to get rid of weback 4.

   I will need to update this PR or https://github.com/hashintel/hash/pull/308 depending on which one gets merged. CI won’t let me forget 😅

## 🐾 Next steps

- Upgrade webpack to 5 in blocks – [Asana task](https://app.asana.com/0/0/1201836863408318/f) _(internal)_

- Sense-check Next.js plugins in Þ repo  – [Asana task](https://app.asana.com/0/0/1201815866759884/f) _(internal)_
  (related to a recent Next.js bump there - https://github.com/blockprotocol/blockprotocol/pull/216)


## 🔍 What does this change?

- Next 12.1 🚀

- swc instead of babel for faster dev experience
  - I removed `packages/hash/frontend/.babelrc`
  - I updated `packages/hash/frontend/jest.config.js` to use `const nextJest = require("next/jest")` (cool stuff!)

- `@svgr/webpack` instead of `next-images`
   - inspired by https://dev.to/dolearning/importing-svgs-to-next-js-nna#svgr
   - `<Svg />` is not `any` any more
   - no type clashes for static PNG imports etc (which are perfectly handled by Next itself)

- no more `FixMeLater` type (alias to `any`) – was used in two cases

- no `prismjs` dependency in frontend app – syntax is still highlighted:
  <img width="676" alt="Screenshot 2022-02-16 at 21 21 18" src="https://user-images.githubusercontent.com/608862/154362821-c487ff46-8243-4ffc-9ffa-5263bf527fbe.png">
   Looks like we added this dependency before creating the code block. Removing prismjs makes sense because syntax highlighting is the responsibility of a block, not the embedding application. Not removing it means we have to use babel instead of swc, which is slower.

### 📜 Does this require a change to the docs?

No

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1201707629991380/1201839331852112/f) (_internal_)

## 🛡 What tests cover this?

- CI (linting, Playright, unit)
- Local manual tests
- Cross-check on the Þ side:
  - https://github.com/blockprotocol/blockprotocol/pull/224
  - https://blockprotocol-2guzghk1k-hashintel.vercel.app/hub

## ❓ How to test this?

1.  Checkout the branch and install deps
1.  Run the app
1.  Make sure all images are loaded as before
1.  Ensure that syntax in code blocks highlighting works as before
